text/0022-pibd-deployment.md

```diff
@@ -0,0 +1,444 @@
+
+Note that if the horizon header changes (i.e. the 12-hour horizon window 'rolls over') while the node is in the process of syncing, there should be no need to invalidate the partial MMRs that have already been downloaded. The node should:
+
+1. Invalidate all previous segment responses whose inclusion proofs don't target the newly computed horizon and re-calculate the list of required segments based on the new horizon
+1. Re-request output bitmap segments and recreate a new output bitmap set
+1. Continue requesting segments based on the new data and the size of the partial MMRs that have already been downloaded.
+
+Note that if the horizon header changes, outputs in the output bitmap set that were previously unspent could potentially become spent. Before validating MMR roots and UTXOs, the node must compare the entire bitmap set against local MMR storage to ensure outputs have been properly marked as spent and/or pruned locally see [Post-PIBD Validation](#post-pibd-validation).
+
+#### Determining segment heights
+
+The [PIBD Messages RFC](https://github.com/mimblewimble/grin-rfcs/blob/master/text/0020-pibd-messages.md) specifies the range of valid segment heights for each MMR, and it is theoretically possible to dynamically adjust segment sizes (and therefore request size) based on network conditions or a host of other factors. However, for simplicity the core implementation currently selects default sizes for each type of segment and uses them throughout for all calculations. These default values are:
+
+```
+BITMAP_SEGMENT_HEIGHT = 9
+OUTPUT_SEGMENT_HEIGHT = 11
+RANGEPROOF_SEGMENT_HEIGHT = 11
+KERNEL_SEGMENT_HEIGHT = 11
+```
+The processes outlined in later sections will assume constant segment heights, but implementations may choose to adjust these heights and/or perform segment size calculations dynamically.
+
+To determine the number of segments `n` of height `h` required for an MMR of a given size `s`, including a possible incomplete last segment, the following calculation can be used:
+
+```
+ceiling(n_leaves(s) / pow(2, h))
+```
+
+where `n_leaves` is the function that calculates the number of leaves in a pmmr of a given size.
+
+Alternatively, a node can attempt to blindly request segments sequentually until one is returned that is not a 'full' segment, however this strategy may lead to unnecessary requests and difficulties determining how much of the PIBD process remains.
+
+#### Peer Selection
+
+The syncing node should select a different peer for each segment request according to the following strategy:
+
+1. Begin with a list of all known peers
+1. Filter this list to only include peers advertising the maximum known difficulty on the network 
+1. Further filter this list to only include peers advertising the [PIBD_HIST_1](#capabilities-flag) flag
```

Comment:
I think the filtering on PIBD_HIST_1 should precede the more complicated filtering on being at max diff...
I'd further consider merging it with the first step as: Begin with a list of all peers advertising the 
 PIBD_HIST_1 capabilities-flag.

Reply:
Yes, that makes sense as an optimisation, updated